src/channel/mpmc.rs

```diff
@@ -107,13 +107,27 @@ where
         wake_send_waiters(send_waiters);
     }
 
+    /// Attempt to send a value without blocking.
+    fn try_send(&mut self, value: T) -> Result<(), ChannelSendError<T>> {
+        if self.buffer.can_push() && !self.is_closed {
```

Comment:
The method is correct. However there is a bit of a downside that this will never work with unbuffered channels. There the transaction happens when the receiver polls the channel and transfers objects directly from blocked senders.

I'm wondering whether it would be too surpising if `try_send` never works for unbuffered channels - while all other methods are just fine.  At least it's not easy to understand without being aware about the internals.

Thinking about how it works in Go: There you can do `try_send` by doing a  `select` and waiting In Go you can `try_send` on an unbuffered channel too with:
```go
select {
    case channel <- value:
default:
}
```` 

The equivalent of this actually also doesn't work with this implementation, since the initial send  will never be ready.
Note: we also can't handover the value directly to a registered receive waiter, since that one might cancel it's transaction and the value would get lost. So maybe documenting the restriction for unbuffered channels would be the way to go.

Reply:
> I'm wondering whether it would be too surpising if try_send never works for unbuffered channels

Maybe we could add a try_send method only if the `ArrayBuf` implements a marker `Buffered` trait?

> we also can't handover the value directly to a registered receive waiter, since that one might cancel it's transaction and the value would get lost.

Will do.

Reply:
We could panic `try_send` if the underlying's buffer capacity is zero. This way its obvious.

Reply:
I like the marker trait idea, but I think it won't work for `HeapRingBuf` - where the size is configured at runtime. `panic!` might be the best compromise. Then it can at least be detected early in the development cycle.

Reply:
Do you think a `debug_assert` + documentation would make sense?

Reply:
Sounds good to me. It should still do something reasonable in release mode (aka: not crash). But I think your implementation does this by returning that it can't push.

Reply:
Yes it should return and error everytime since `self.buffer.can_push` is always false for a unbuffered channel (unless the user decides to do some weird impl).